stacks-signer/src/v0/signer_state.rs

```diff
@@ -38,6 +40,205 @@ use crate::signerdb::SignerDb;
+        let mut protocol_versions = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let entry = protocol_versions
+                .entry(update.local_supported_signer_protocol_version)
+                .or_insert_with(|| 0);
+            *entry += weight;
+            if *entry >= self.total_weight * 7 / 10 {
+                return Some(update.local_supported_signer_protocol_version);
+            }
+        }
+        None
+    }
+
+    /// Determine what the global burn view is if there is one
+    ///
+    /// NOTE: do not call this function unless the evaluator has already been updated with the current local state
+    pub fn determine_global_burn_view(&self) -> Option<(ConsensusHash, u64)> {
+        let mut burn_blocks = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let StateMachineUpdateContent::V0 {
+                burn_block,
+                burn_block_height,
+                ..
+            } = update.content;
+
+            let entry = burn_blocks.entry(burn_block).or_insert_with(|| 0);
+            *entry += weight;
+            if *entry >= self.total_weight * 7 / 10 {
+                return Some((burn_block, burn_block_height));
+            }
+        }
+        None
+    }
+
+    /// Check if there is an agreed upon global current miner viewpoint
+    ///
+    /// NOTE: do not call this function unless the evaluator has already been updated with the current local state
+    pub fn determine_global_state(&self) -> Option<SignerStateMachine> {
+        let active_signer_protocol_version = self.determine_active_signer_protocol_version()?;
+        let (global_burn_block, _) = self.determine_global_burn_view()?;
+        let mut miner_views = HashMap::new();
+        for (address, update) in &self.address_updates {
+            let Some(weight) = self.address_weights.get(address) else {
+                continue;
+            };
+            let StateMachineUpdateContent::V0 {
+                burn_block,
+                burn_block_height,
+                current_miner,
+                ..
+            } = &update.content;
+
+            if *burn_block != global_burn_block {
+                continue;
+            }
+
+            let entry = miner_views.entry(current_miner).or_insert_with(|| 0);
+            *entry += weight;
+            if *entry >= self.total_weight * 7 / 10 {
+                return Some(SignerStateMachine {
+                    burn_block: *burn_block,
+                    burn_block_height: *burn_block_height,
+                    current_miner: current_miner.into(),
+                    active_signer_protocol_version,
+                });
+            }
```

Comment:
I think we could get away with something like this:

```suggestion
            let StateMachineUpdateContent::V0 {
                burn_block,
                burn_block_height,
                current_miner,
                ..
            } = &update.content;

            let state_machine = SignerStateMachine {
                    burn_block: *burn_block,
                    burn_block_height: *burn_block_height,
                    current_miner: current_miner.into(),
                    active_signer_protocol_version,
            };

            let entry = state_views.entry(state_machine).or_insert_with(|| 0);
            *entry += weight;
            if *entry >= self.total_weight * 7 / 10 {
                return Some(state_machine);
            }
```

Where we just look for agreement across the whole state machine struct.